go.mod

```diff
@@ -3,12 +3,16 @@ module sigs.k8s.io/cli-utils
 go 1.16
 
 require (
+	github.com/google/go-cmp v0.5.6
 	github.com/google/uuid v1.2.0
+	github.com/mikefarah/yq/v4 v4.12.1
 	github.com/onsi/ginkgo v1.16.2
 	github.com/onsi/gomega v1.12.0
+	github.com/pkg/errors v0.9.1
 	github.com/spf13/cobra v1.1.3
 	github.com/stretchr/testify v1.7.0
-	gopkg.in/yaml.v3 v3.0.0-20200615113413-eeeca48fe776
+	gopkg.in/op/go-logging.v1 v1.0.0-20160211212156-b2cb9fa56473
```

Comment:
Please don't add any logging dependencies to this library. As a consumer, I'd like the decision of what logger implementation to use to be left to me.

Reply:
go-logging is used by the yq impl that's imported.

I added a logutil package to allow configuring go-logging to use klog as a backend.

Reply:
Oh, I see.

https://github.com/op/go-logging/blob/b2cb9fa56473e98db8caba80237377e83fe44db5/backend.go#L7-L28 the package field that `logging.SetBackend()` sets is not protected by a mutex so this method is only safe to call from `init()` or `main()` before any package has a chance to start any goroutines. Otherwise there is a risk of a data race.

I suggest to leave it out and let consumers do something about it if they want to. But whatever they do, it should be in `main()` or in an `init()`.

For tests (if we really need to set the backend for them?), it should be possible to set a backend in [`TestMain()`](https://pkg.go.dev/testing#hdr-Main) - that should be safe. But I also suggest to leave it out if possible to avoid making this logging library a direct dependency. Having it as a transitive dependency is unavoidable if we need to use this library, but it doesn't have to be a direct one. WDYT?

Reply:
Forgive my ignorance, but if the defaultBackend is only set once by SetBackend and not accessed beforehand, why does it need a mutex? What's the race?

> For tests (if we really need to set the backend for them?), it should be possible to set a backend in TestMain()

Yes, I could add a TestMain for the test packages that execute yq code. It's only run once per package, not per test, but that should be fine as long as we're not testing the logs themselves.

> Having it as a transitive dependency is unavoidable if we need to use this library, but it doesn't have to be a direct one. WDYT?

If we decide to accept this yq dependency, then my assumption is that every client that uses cli-utils will need to wrap go-logging with klog, otherwise the logging defaults to trace level on stderr, with a default format, and yq logs extensively. This assumes all cl-utils clients use the apply package and want to support apply-time-mutation, but I suspect that's a 100% overlap.

I also expect almost all cli-utils clients use klog for logging, because it's the Kubernetes standard logger. But I'm willing to be contradicted on that point. I don't know of any full list of usages.

I'm not trying to be stickler, here. I totally understand not wanting to add a new log dependency. I want to do the right thing. But if on of cli-utils's dependencies starts logging to another logger, I would expect that clients will want to be warned in release notes and provided with a workaround.

I'm willing to consider moving the logutil to its own package tho, and depending on that for tests and recommending it to clients. But making a new OSS repo means I'll need to go through a corporate process for approval. So I wanna make sure it's the best option first.

Reply:
> but if the defaultBackend is only set once by SetBackend and not accessed beforehand, why does it need a mutex? What's the race?

We (from the point of view of this library) don't know if anything else in the user's program is using the same logging library. There may be read accesses to that field from other goroutines. To make it safe, there must be memory ordering guarantees between the write and any subsequent reads. To have such guarantees, the library user must ensure the write happened before those goroutines are started. The only way to achieve this reliably is to perform the write in `main()` or `init()` - before any goroutines are started. Does it explain my concern? I'm not saying we need to wrap it with a mutex - only the logging library author could do that, but they chose not to do that (which is fine, but it puts the above constraints on consumers).

I now see why you've done what you've done, it makes sense to me. I suggest to export the `klogGoLoggingBackend` type and remove the `WrapGoLogging()` function. I don't think providing it is very valuable, I think having the `logging.SetBackend()` call in the user's program would be better because it's more explicit. Perhaps we could add a note somewhere that it must be called.

> I'm willing to consider moving the logutil to its own package tho

(I think you mean a separate repo/module) I think it's not worth it. Having it as a separate package (not a repo) is fine in my opinion.

Reply:
Exporting the backend struct seems reasonable. Removing the WrapGoLogging function won't remove the go-logging dependency, but I can move `logging.SetBackend(klogGoLoggingBackend{})` to the struct comment and call ti directly from TestMain, since it's a one-liner.

Reply:
Ok. I've done a bit more than that and made KlogGoLoggingBackend more flexible with configurable log and verbosity mappings. Added comments to explain why and how to use.

I've also validated it works as expected by using it in the kyq tests with TestMain.

```
go test sigs.k8s.io/cli-utils/pkg/kyq -v -args -v=5
```

Reply:
👍